airflow/hooks/hdfs_hook.py

```diff
@@ -17,85 +17,95 @@
-    :type autoconfig: bool
-    """
-    def __init__(self, hdfs_conn_id='hdfs_default', proxy_user=None,
-                 autoconfig=False):
-        if not snakebite_imported:
-            raise ImportError(
-                'This HDFSHook implementation requires snakebite, but '
-                'snakebite is not compatible with Python 3 '
-                '(as of August 2015). Please use Python 2 if you require '
-                'this hook  -- or help by submitting a PR!')
         self.hdfs_conn_id = hdfs_conn_id
-        self.proxy_user = proxy_user
-        self.autoconfig = autoconfig
+        self._autoconf = autoconf
+
+        self._conn = None
 
     def get_conn(self):
-        """
-        Returns a snakebite HDFSClient object.
-        """
-        # When using HAClient, proxy_user must be the same, so is ok to always
-        # take the first.
-        effective_user = self.proxy_user
-        autoconfig = self.autoconfig
-        use_sasl = configuration.conf.get('core', 'security') == 'kerberos'
-
-        try:
-            connections = self.get_connections(self.hdfs_conn_id)
-
-            if not effective_user:
-                effective_user = connections[0].login
-            if not autoconfig:
-                autoconfig = connections[0].extra_dejson.get('autoconfig',
-                                                             False)
-            hdfs_namenode_principal = connections[0].extra_dejson.get(
-                'hdfs_namenode_principal')
-        except AirflowException:
-            if not autoconfig:
-                raise
-
-        if autoconfig:
-            # will read config info from $HADOOP_HOME conf files
-            client = AutoConfigClient(effective_user=effective_user,
-                                      use_sasl=use_sasl)
-        elif len(connections) == 1:
-            client = Client(connections[0].host, connections[0].port,
-                            effective_user=effective_user, use_sasl=use_sasl,
-                            hdfs_namenode_principal=hdfs_namenode_principal)
-        elif len(connections) > 1:
-            nn = [Namenode(conn.host, conn.port) for conn in connections]
-            client = HAClient(nn, effective_user=effective_user,
-                              use_sasl=use_sasl,
-                              hdfs_namenode_principal=hdfs_namenode_principal)
-        else:
-            raise HDFSHookException("conn_id doesn't exist in the repository "
-                                    "and autoconfig is not specified")
-
-        return client
+        if self._conn is None:
+            if self.hdfs_conn_id is None:
+                self._conn = hdfs3.HDFileSystem(autoconf=self._autoconf)
+            else:
+                params = self.get_connection(self.hdfs_conn_id)
+
+                # Extract hadoop parameters from extra.
+                hdfs_pars = params.extra_dejson.get('pars', {})
```

Comment:
Maybe `hfds_pars` could be augmented like so

```python
if configuration.conf.get("core", "security") == "kerberos":
    hdfs_pars["hadoop.security.authentication"] = "kerberos"
```

to make existing code not break?

Reply:
Good idea. Should the airflow configuration override whatever is set in hdfs_pars? (I.e., should we keep the value for `hadoop.security.authentication` if it is already set in `pars` or should we override it?)